### ISSUE TITLE
Implement supply functions for SMT's

### ISSUE THREAD

**Issue by theoreticalbts** [steemit/steem#1625]
We need to be able to adjust SMT balances and track SMT's in supply accounting.  EDIT:  This work is complex and involves several moving parts and subtle issues.  So let's start things off with a big-picture discussion of what parts of Steem need to change to support arbitrary asset balances, and file separate issues for specific tasks as necessary.

**Comment by theoreticalbts** [comment]
Let's start with a historical review.  Originally Steem was not designed to have multiple asset types, so an account's STEEM, SBD and VESTS balances exist as specific fields on the account object.  Likewise, the savings balances exist as specific fields.

So I started to write down questions, it seems clear that maybe each of these questions should be its own issue.  So I will put checkboxes here, to be checked as each issue is closed.  I will also create issues for each item, and edit this post to include issue numbers of each issue.

- [ ] It is clear that SMT balances must be implemented as a separate object indexed by `account_name`, `smt_symbol` fields.
- [ ] Should STEEM, SBD and VESTS balances continue to be fields on the account object, or should they be refactored to be separate objects as well?
- [ ] Currently SBD interest processes whenever an SBD balance changes.  Should this behavior continue?
- [ ] Should we refactor SBD balance into a struct like this?  It would allow unification of handling of SBD balance inside / outside savings.

```
struct account_sbd_balance
   asset             sbd_balance;
   uint128_t         sbd_seconds;
   time_point_sec    sbd_seconds_last_update;
   time_point_sec    sbd_last_interest_payment;
};
```

- [ ] Is the `price` structure and market handling system able to handle multiple different kinds of orders?  I think the answer is "yes" since much of the code is borrowed from BitShares, which can.  But I'm not sure, and it certainly needs tested.
- [ ] Should we move overflow checks to balance adjustment functions?  It seems like this is a no-brainer refactoring to make it harder to create overflow bugs.
- [ ] I'm thinking about writing a compile-time bounded integer type which automatically bounds checks the result of any arithmetic operation.  Should we do this?
- [ ] What is the global upper bound on the number of satoshis of an asset that can exist?
  Is it `STEEMIT_MAX_SHARE_SUPPLY`?   Is it `numeric_limits<uint64_t>::max()`?  Do we need to define a different constant?  Is it unbounded so that we must switch to arbitrary precision arithmetic at some point?  How do we handle occasions when per-block processing would cause the supply bound to be exceeded?  Does the current code check all possible places overflow condition can occur?  Do unit tests exercise any / all possible overflow conditions?  Do we need to adapt Bitcoin's design decision that the maximum number of satoshis must be smaller than the mantissa of a floating-point number because of JavaScript clients?  Related: #1640
- [ ] According to the specification, SMT's don't have savings functionality, but we didn't think very hard about it.  Should SMT's have savings functionality?
- [ ] Supply accounting functions (`adjust_supply()` and `validate_invariants()`) need to be adapted to check SMT supply.
- [ ] Is `validate_invariants()` run after every unit test?  Is there a way to run `validate_invariants()` on the main network?  At one time every node would run `validate_invariants()` every block, but this was abandoned once usage grew to the point where `validate_invariants()` took more than a few hundred ms.  Do we have any automated process in place to make sure `validate_invariants()` runs regularly on the main network?

**Comment by mvandeberg** [comment]
> I'm thinking about writing a compile-time bounded integer type which automatically bounds checks the result of any arithmetic operation. Should we do this?

Assets are backed by `share_type` which is already bounds checked.

``` c++
typedef safe<int64_t>               share_type;
```

**Comment by mvandeberg** [comment]
With regard to where STEEM, SBD, and VESTS should live, at a minimum VESTS need to be on the account object. Vesting balances impact witness votes which impact the witness schedule. Thus, they are one of the most fundamental fields in the blockchain. This is most pertinent to the modularization of the blockchain.

STEEM and SBD can exist outside of this module because `transfer_to_vesting_operation` is self describing. The module that contains the STEEM balance can read from the witness schedule/vesting module to determine how much STEEM is withdrawn when a power down occurs.

If we are going to make such a refactor, we should ensure that the refactor supports the future architecture of STEEM.

**Comment by theoreticalbts** [comment]
AFAIK `safe<t>` has fixed lower and upper bounds of `numeric_limits<T>::min()` and `numeric_limits<T>::max()`.  I am thinking that the following new features should be added:

- The lower/upper bounds should be template parameters instead.
- `asset` / `share_type` fields used in objects and operations should have lower limit of 0 and upper limit of `STEEM_MAX_SHARE_SUPPLY`
- Serialization / deserialization does bounds checks
- Arithmetic on assets results in type with template parameters with new different bounds.  That is, `safe<A,B> + safe<C,D> = safe<A+C, B+D>` since if `A <= x <= B, C <= y <= D` then `A+C <= x+y <= B+D`.  Likewise `safe<A,B> - safe<C,D> = safe<A-D, B-C>` since if `A <= x <= B, C <= y <= D` then `A-D <= x-y <= B-C`.
- You can explicitly convert `safe<A,B>` to `safe<C,D>` which adds runtime bounds checks.

Basically the idea's that by default, the compiler will enforce that code doesn't accept unreasonable values, i.e. negative or greater than `STEEM_MAX_SHARE_SUPPLY`.  Code that manipulates amounts with arithmetic may generate temporary variables that are out-of-bounds, but the compiler will force you to add an explicit runtime bounds check in order to store such a temporary in an object field.

**Comment by mvandeberg** [comment]
Having an upper bound of `STEEM_MAX_SHARE_SUPPLY` is problematic as demonstrated in #1640 

We would need to make the asset type also templated and have well defined semantics when creating prices between assets with different bounds.

For the record, I do not believe that negates the changes in #1641

**Comment by mvandeberg** [comment]
I propose the templated asset type called bounded_asset and templated price type (inferred template parameters for price to preserve existing code) and typedef asset as bounded_asset< 0, STEEM_MAX_SHARE_SUPPLY > and vesting_asset as bounded_asset<> (default to numerical limits).

**Comment by theoreticalbts** [comment]
> Having an upper bound of STEEM_MAX_SHARE_SUPPLY is problematic as demonstrated in #1640

Agreed.  We should create a new upper bound called `STEEM_MAX_SATOSHIS` or something.

> We would need to make the asset type also templated and have well defined semantics when creating prices between assets with different bounds.

I think that it suffices to say `price` can only be constructed by assets with the global bounds, an `asset` with any other bounds will require an explicit conversion.

**Comment by youkaicountry** [comment]
Closing as outdated.
